### ISSUE TITLE
Reviewing PRs

### ISSUE THREAD

**Issue by normanb** [locustio/locust#355]
what is the process for reviewing and accepting a PR with locust? I see a number of open PRs (including mine) that have been open for some time. I just trying to get an idea of when PRs will be merged or rejected into master or if the project is considered stable and not accepting PRs.

**Comment by molsky** [comment]
@heyman @Jahaja  @cgbystrom 

What is the current situation? There are a lot of PRs that would improve things from test automation view (fix for console ouputs and way to get CSV files when running Locust scripts headless).

**Comment by Chumper** [comment]
:+1: looks like this project needs more maintainer. Nothing happened since September and i would love to see some PR merged in.

**Comment by pmdarrow** [comment]
I'm willing to be a maintainer but the project owners won't seem to reply to any messages.

**Comment by molsky** [comment]
Current situations seems pretty bad, has anyone managed to contact one of those project maintainers?

**Comment by justiniso** [comment]
I just tried to reach them on twitter: https://twitter.com/JustinIso/status/684469178808537088

**Comment by pmdarrow** [comment]
I emailed & tweeted all of the present & past maintainers but haven't received a reply from any of them. I've decided to fork the project and will be releasing it under a different name. Stay tuned...

**Comment by sivabudh** [comment]
@pmdarrow Kudos!

**Comment by peterbe** [comment]
I strongly encourage the original locust team to consider Jazzband
https://jazzband.co/news/2015/12/17/launching-jazzband

**Comment by mthurlin** [comment]
I'll ping the maintainers IRL, and see what we can do to make the project more actively maintained

**Comment by justiniso** [comment]
Thanks @mthurlin @pmdarrow happy to help in any way I can and willing to contribute as a maintainer if extra hands are needed.

**Comment by pmdarrow** [comment]
@mthurlin you know the maintainers? Let them know that there are several people willing to maintain the project if they're no longer interested (myself included). Would love to not have to fork!

**Comment by CCedricYoung** [comment]
I'm new to locustio and also would like to see some of the PR's, like #371, get resolved. Is anyone actively maintaining a fork that has this?

**Comment by justiniso** [comment]
@CCedricYoung needless to say I would also like to see #371 merged :wink: But I wouldn't get your hopes up with this being resolved soon. Even if the maintainers returned today, it may be a while before any pull requests can be fully reviewed and merged, or a new version released to pypi. 

Our company does have a fork that incorporates those changes: https://github.com/jwplayer/locust, but I cannot offer any support or documentation beyond the code. At this point we're using it internally and are waiting to see how this situation evolves. Hope it helps.

**Comment by cgbystrom** [comment]
Sorry for delayed answers/merging of PRs. We haven't really been giving Locust the love it needs lately.

I spoke briefly to @heyman and we are more than happy to onboard another maintainer.
Need to talk some more to him, then I'll make sure to get back to you guys.

**Comment by imreACTmd** [comment]
+1 for @pmdarrow as another maintainer

**Comment by cgoldberg** [comment]
> we are more than happy to onboard another maintainer.

would you consider giving the commit bit to 2-3 new maintainers?  There are currently 100 queued up issues, 48 open PR's, and several forks maintaining additional features and code.  That might be a lot of work for 1 more to undo the current logjam.  (and there is some great stuff in there btw)

At the beginning they can just deal with more mundane PR's (bug fixes, etc), and any new features or architectural designs can always be given to you.  And i think we need to up the internal automated  tests significantly.... a mix of unit and functional tests with decent coverage.  This will become critical once the Python3 support PR lands,  `tox`and `detox` would be a really good fit here... for getting fast local developer runnable multi-version testing going.  It works nicely in combination with the existing travic-ci setup (you actually can simplify your travic-ci configuration by just calling `tox` to handle everything)

https://tox.readthedocs.org/en/latest/
 
thoughts on anything?  or any updates on this?

**Comment by molsky** [comment]
@cgbystrom Have you done any decisions yet for new maintainers? It would be great to get this project back to alive.

**Comment by kamikaze** [comment]
c'mon guys, fork it and maintain it then! Also rewrite it by using Python 3.5!

**Comment by pmdarrow** [comment]
@kamikaze there's no need to rewrite it for Python 3, see the PR I've made here: https://github.com/locustio/locust/pull/363. There's also no need to fork if owners allow a few more committers.

**Comment by kamikaze** [comment]
what are we waiting for? Let's fork it! Also writing Python 2 these days is like eating food with hands after toilet without washing them

**Comment by justiniso** [comment]
@kamikaze appreciate the enthusiasm, but it won't do anyone any good to fragment the locust community. Forking should generally be the last resort and the owners have offered to onboard a new maintainer, so we're waiting to see if that happens. 

Also rewriting from scratch would be a needlessly massive effort and would further delay the delivery of features. As @pmdarrow mentioned, https://github.com/locustio/locust/pull/363 ports the existing project to 3 and supports tests for python 3 (which passed at the time). So a rewrite wouldn't provide much benefit.

**Comment by jmelfi** [comment]
any update here? @heyman @cgbystrom It's been over a month and no news that I've found yet here about new maintainers yet.

**Comment by deniscostadsc** [comment]
Hi guys.

I work at ThoughtWorks, one of the companies listed as Locust users on locust website. As a user I don't want to see this project dies.

I'd like to help the project, in any way. What can we do for the project? Could add any of these guys above as maintainers?

Just tell us and the community will take care of the project.

**Comment by pmdarrow** [comment]
@deniscostadsc we've be waiting for almost 2 months for their response. I think it's getting to the point where we need to fork...

**Comment by deniscostadsc** [comment]
Yeah...

I'm using Locust right now, and I'm missing some features. I'd like to PR those features, but it's not gonna happen...

I think to fork is not a problem. If they come back, we can merge again.

**Comment by justiniso** [comment]
:+1: it's definitely time!

@pmdarrow I think you mentioned you started a fork already? If you make it public, I would support it.

**Comment by jmelfi** [comment]
@pmdarrow @deniscostadsc @justiniso I'd be glad to help with this project as our company has been using it just a bit so far.

I have some spare time to work on a fork.  I'd be glad to assist.

**Comment by jinxcat** [comment]
I also definitely support this project not dying, as it has many advantages compared to alternatives. I've used it in the past extensively and I would like to use it in the future as well. If it continues like this though, it will just starve from lack of attention.

I'd be glad to assist any effort in maintaining this project and bringing it up to date to users' needs.

**Comment by smrpr** [comment]
@pmdarrow @deniscostadsc @justiniso @jmelfi @jinxcat let's do it :+1:

**Comment by deniscostadsc** [comment]
How do you guys think we can do this?

**Comment by jmelfi** [comment]
I believe that pmdarrow had a [fork here](https://github.com/pmdarrow/locust) but he hasn't responded here to indicate that the fork there is a good place to start.  

I have an up-to-date [fork here](https://github.com/jmelfi/locust) that i'd willingly to give to an org or add contributers to it so that it could be worked on.

@jinxcat @smrpr @deniscostadsc @justiniso @pmdarrow It seems that we have a good amount of interest here and if we need to hard fork, I say let's get started.  I can add collaborators to my public repo and I can setup waffle.io if we needed to do some management.  Thoughts?

**Comment by pmdarrow** [comment]
I haven't moved forward with the fork since I was still kind of holding out hope that we could get the maintainers to add one or more of us as contributors, but it's becoming clear that likely won't happen. Do you think we should fork under a different name to reduce confusion with this version?

**Comment by jmelfi** [comment]
@pmdarrow I'd fork under a different name, though I'm unsure of what to use.  It would likely help ensure there isn't confusion with this repo and the fork.

I'm open to moving this forward.  I don't have a huge amount of time to devote to this (just some weekend time really).

All these improvements can be done, we need to decide on a repo to use.  I'd gladly open mine up.  I believe pmdarrow's fork could easily be merged into there.  I can change my fork's name once it's decided that we'd use that fork for the **hard fork**.  

@pmdarrow @smrpr @deniscostadsc @justiniso @jinxcat please reply with a request input.  If we desire to use my fork for the **hard fork,** we can work out managing merges, testing automation, and chat.  With the stagnation here, I think acting and moving forward is the best option to keep this project alive.

**Comment by smrpr** [comment]
I'm fine with what @jmelfi is proposing.

**Comment by justiniso** [comment]
Internally, my company has been using the name "cicada" with the emphasis on CI (for continuous integration). Our fork incorporated a couple of the pull requests that make it easier to run locust as part of a Jenkins build: saving results to file from command line, allow --no-web with distributed, and pointing to multiple locustfiles. 

If you can handle the cheesiness, I have the pypi namespace for it.

**Comment by justiniso** [comment]
For the fork, I see a couple of options:

- Create an organization under the new name to host the repo
- My company would be happy to host and support the repo: https://github.com/jwplayer, which we've talked about for our own version and have a long tradition of supporting open-source. 
- Anyone else willing to host with organizational/corporate support

I'm a little hesitant to put it under any one person's personal repo. Aside from purely superficial appearances to keep up legitimacy, it gives all the keys to the castle to one person who could go rogue, introduce malware, or disappear and put us back in this same situation.

**Comment by smrpr** [comment]
I agree on creating not putting it under somebody's personal account. Nicest thing would be the new account/everything but it's more of a hassle than @justiniso 's company hosting it. I honestly don't mind.

**Comment by cgoldberg** [comment]
I'm +1 for a new org, comprised of volunteers from this thread.  Hosting for orgs with public repos is free AFAIK.

> My company would be happy to host and support the repo

What benefits does that support have over a small community of outside contributors?  Would external contributors be able to merge code into master?

**Comment by pmdarrow** [comment]
I can create the org, any name suggestions? I was thinking something like grasshopper, akris (greek for locust), swarm, ... I'm bad at this :laughing:

**Comment by justiniso** [comment]
@cgoldberg mostly moral support and publicity to our developer community. Absolutely the maintainers would be able to commit to master. We're not looking to own the project, just publicly back it. But just offering it as an option to make things easier. 

@pmdarrow see my comment above about "cicada" or ("CI-cada") :wink:

**Comment by zvxr** [comment]
+1 for cicada

For merging into master, would there be a review process? Guide? Ensuring the Travis tests pass would be a great start.

**Comment by pmdarrow** [comment]
Whoops, didn't see that @justiniso. I like that, works for me. github.com/cicada is taken, what about cicadaorg? cicadatool? or @therealcicada? :joy: Any ideas?

**Comment by smrpr** [comment]
+1 for cicadaorg

**Comment by justiniso** [comment]
cicadaCI? cicada-load / cicada-perf if we want to be more explicit. 

cicada.ci is available but cicada.io is taken :(

**Comment by pmdarrow** [comment]
Only thing about using "CI" is it sounds like it's a continuous intergration server, when it's not really. I like cicada-load the best out of those.

**Comment by jmelfi** [comment]
sounds like quite a bit of progress going here.  I agree that we should have it in an org and not a personal.

@zvxr I would expect tests to pass and new test to be written/modified as new features are added.

@pmdarrow I agree that "CI" is probably not a great thing to add to this.  +1 for akris if that's not taken.  I suck with names so not a great person to ask either.

**Comment by jinxcat** [comment]
+1 for akris

I would strongly suggest the creation of a new organisation, not directly affiliated with any particular company, with a (high? 5+) number of equal rights owners. That way we will not run into the same problems of "vanished" owers.

Another issue that must be discussed is a set of "rules"/guidelines we need to follow. Examples: no PR accepted without tests, no PR accepted without 2 or more maintainers giving consent, discussion on issues and not PRs, and so on.

**Comment by justiniso** [comment]
I created a slack team to start figuring out the details. I'll set up a signup page and post it here.

In the meantime, if you're interested in contributing early on, shoot me an email and I'll manually invite you -->  inviteplease@justiniso.com

**Comment by cgoldberg** [comment]
I just added a logo to the cicada-load org (it's a cicada).
https://github.com/cicada-load

I dig it... does anyone object? :)
btw, it is CC licensed

**Comment by pmdarrow** [comment]
I dig it! Let's move discussion over there so we don't hijack this repo any more. https://github.com/cicada-load/cicada if anyone is interested in an maintained version of locust!

**Comment by ajazzarello** [comment]
Awesome thank you!



Sent via the Samsung Galaxy Note® 4, an AT&T 4G LTE smartphone


-------- Original message --------

**Comment by cgoldberg** [comment]
@pmdarrow i just started adding issues for new stuff I want to discuss.  we can discuss there

**Comment by cgbystrom** [comment]
Sorry again for the lack of updates here from my side / other maintainers.
There have been other events in life getting in the way for me putting more time into this.

I see you have proceeded with forking Locust, a very understandable decision given recent developments in this thread. This is of course sad to see but we (as maintainers) have nothing to blame but ourselves for this situation.

The previously discussed push rights to the repo have given to @pmdarrow, @cgoldberg and @justiniso. I realize this may be too late and I can only urge you to continue in the `locustio/locust` repo, as I think it would be in the best interest of the community. 

But given the nature of open-source and GitHub, you should of course proceed in a way you best see fit.

**Comment by arthurlogilab** [comment]
I don't get it... where did the fork go ? all the cicada and cicada-load URLs end up on a 404.

**Comment by Chumper** [comment]
@arthurlogilab They added some of the folks as maintainer so we decided to go along with this repo again.

**Comment by pmdarrow** [comment]
PRs are now being reviewed, so closing this for now!

**Comment by di** [comment]
Really glad to see this project getting some new maintainers! Thanks @pmdarrow, @cgoldberg and @justiniso for stepping up. I'd be glad to join the team as well if more help is needed.
